File: python_ws/0383_ransom_note.py
# time: 160ms, beat 13.39%
# RAM: 12.8 MB, beat 99.92%
class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        for i in ransomNote:
            if ransomNote.count(i) > magazine.count(i):
                return False

        return True

# time: 56ms, beat 54.45%
# RAM: 13.1MB, beat 72.77%
class Solution2(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        dict = {}
        for i in magazine:
            dict[i] = dict.get(i, 0) + 1
        for j in ransomNote:
            dict[j] = dict.get(j, 0) - 1
            if dict[j] < 0:
                return False

        return True
# use array to store the frequency of character saves more space, since 
# it will not need to store the key
# time: 44ms, beat 79.43%
# RAM: 13MB, beat 91.4%
class Solution3(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        array = [0] * 26
        # ord function to transform the char to ascii
        for i in magazine:
            array[ord(i) - ord('a')] += 1
        for j in ransomNote:
            array[ord(j) - ord('a')] -= 1
            if array[ord(j) - ord('a')] < 0:
                return False

        return True

if __name__ == "__main__":
    solution = Solution()

    ransomNote = "a"
    magazine = "b"
    print(solution.canConstruct(ransomNote, magazine))

    ransomNote = "a"
    magazine = "ba"
    print(solution.canConstruct(ransomNote, magazine))

    ransomNote = "ab"
    magazine = "abbb"
    print(solution.canConstruct(ransomNote, magazine))

    ransomNote = "aab"
    magazine = "baa"
    print(solution.canConstruct(ransomNote, magazine))

File: cpp_ws/src/0383_ransom_note.cpp
#include "helper.cpp" 
#include <algorithm>

// time: 260ms, beat 5.12%
// RAM: 8.5MB, beat 68.17%
class Solution {
public:
    bool canConstruct(std::string ransomNote, std::string magazine) {
        for (char i : ransomNote){
            if (std::count(ransomNote.begin(), ransomNote.end(), i) > std::count(magazine.begin(), magazine.end(), i) )
                return false;
        }
    return true;
    }
};

// time: 260ms, beat 5.12%
// RAM: 8.5MB, beat 68.17%
class Solution2 {
public:
    bool canConstruct(std::string ransomNote, std::string magazine){
        int hash[26] = {0}; // use array with size 26 to represent hash table
        std::cout << hash[0] << std::endl;
        for(char i : magazine){
            hash[i-'a'] += 1;
        }
        for(char j : ransomNote ){
            hash[j-'a'] -= 1;
            if(hash[j-'a'] < 0)
                return false;
        }
        return true;
    }
};

int main(){
    Solution2 solution;

    std::string ransomNote1 = "a";
    std::string magazine1 = "b";
    std::cout << solution.canConstruct(ransomNote1, magazine1) << std::endl;
    
    std::string ransomNote2 = "a";
    std::string magazine2 = "ba";
    std::cout << solution.canConstruct(ransomNote2, magazine2) << std::endl;

    std::string ransomNote3 = "ab";
    std::string magazine3 = "abbb";
    std::cout << solution.canConstruct(ransomNote3, magazine3) << std::endl;

    std::string ransomNote4 = "aab";
    std::string magazine4 = "baa";
    std::cout << solution.canConstruct(ransomNote4, magazine4) << std::endl;
}